### PR TITLE
kube-state-metrics: init at 2.12.0

### PR DESCRIPTION
**Pull request by eskytthe**
## Description of changes
This add the k8s kube-state-metrics server. It is used to generating metrics from Kubernetes API objects such as deployments, nodes and pods. It is one of the main sources used to get prometheus metrics out of a k8s cluster.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by estka** [comment]
Result of `nixpkgs-review pr 314208` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kube-state-metrics</li>
  </ul>
</details>

File: pkgs/by-name/ku/kube-state-metrics/package.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    sha256 = "sha256-EBSlufgzT4zgxTO5B0/RHHJJ2GgtU3U9ypd1QjC3knE=";
+  };
+
+  excludedPackages = [ "./tools" ];
+
+  preCheck = ''
+    rm -r tests/e2e
+  '';
+
+  meta = with lib; {
```
Comment (estka) [inline pkgs/by-name/ku/kube-state-metrics/package.nix:27]: Please remove `with lib;`

File: pkgs/by-name/ku/kube-state-metrics/package.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    owner = "kubernetes";
+    repo = "kube-state-metrics";
+    rev = "v${version}";
+    sha256 = "sha256-EBSlufgzT4zgxTO5B0/RHHJJ2GgtU3U9ypd1QjC3knE=";
+  };
+
+  excludedPackages = [ "./tools" ];
+
+  preCheck = ''
+    rm -r tests/e2e
```
Comment (SebTM) [inline pkgs/by-name/ku/kube-state-metrics/package.nix:24]: Could you do `doCheck = false;` instead if there are no other tests (had a short look) or open an issue to provide a command to run only the not e2e-tests and change the command in checkPhase to that if solved upstream?

Comment (eskytthe) [reply pkgs/by-name/ku/kube-state-metrics/package.nix]: Thanks @SebTM - so like this? Did check https://github.com/kubernetes/kube-state-metrics/tree/main/tests/e2e where there seems to be several small tests based on some kind. But yes it does not kind of work in our NixOS environment. Maybe I will later try to contact upstream. But for now this work for us. We have used kube-state-metrics for > 4 years now so about time time to upstream :-).

**Review by SebTM — CHANGES_REQUESTED** [review]
Had another look and checked out the code - stdenv is unused, setting doCheck can be avoided by excludePackages so if they have non e2e-tests or decide to add some they will pop-up.

```nix
{
  lib,
  buildGoModule,
  fetchFromGitHub,
}:

buildGoModule rec {
  pname = "kube-state-metrics";

  version = "2.12.0";
  vendorHash = "sha256-A/oiFEWM3TOtk4V3xpehJ+5oKSyzYKZHVlp3wjykFRA=";

  src = fetchFromGitHub {
    owner = "kubernetes";
    repo = "kube-state-metrics";
    rev = "v${version}";
    sha256 = "sha256-EBSlufgzT4zgxTO5B0/RHHJJ2GgtU3U9ypd1QjC3knE=";
  };

  excludedPackages = [
    "./tests/e2e"
    "./tools"
  ];

  meta = {
    homepage = "https://github.com/kubernetes/kube-state-metrics";
    description = "Add-on agent to generate and expose k8s cluster-level metrics";
    license = lib.licenses.asl20;
    maintainers = [ lib.maintainers.eskytthe ];
    platforms = lib.platforms.unix;
  };
}
```

lastly I applied `nixfmt-rfc-style` on it 😃

**Comment by eskytthe** [comment]
Thanks again @SebTM , your suggestions are appreciated! :-). Please ping me if I should squash the commits, or still miss something.

**Review by SebTM — APPROVED** [review]
LGTM 👍🏻 

Result of `nixpkgs-review pr 314208` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kube-state-metrics</li>
  </ul>
</details>

Please squash if possible - yes ✌🏻

File: pkgs/by-name/ku/kube-state-metrics/package.nix [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  version = "2.12.0";
+  vendorHash = "sha256-A/oiFEWM3TOtk4V3xpehJ+5oKSyzYKZHVlp3wjykFRA=";
+
+  src = fetchFromGitHub {
+    owner = "kubernetes";
+    repo = "kube-state-metrics";
+    rev = "v${version}";
+    sha256 = "sha256-EBSlufgzT4zgxTO5B0/RHHJJ2GgtU3U9ypd1QjC3knE=";
+  };
```
Comment (SuperSandro2000) [inline pkgs/by-name/ku/kube-state-metrics/package.nix:18]:
```suggestion
  version = "2.12.0";

  src = fetchFromGitHub {
    owner = "kubernetes";
    repo = "kube-state-metrics";
    rev = "v${version}";
    hash = "sha256-EBSlufgzT4zgxTO5B0/RHHJJ2GgtU3U9ypd1QjC3knE=";
  };

  vendorHash = "sha256-A/oiFEWM3TOtk4V3xpehJ+5oKSyzYKZHVlp3wjykFRA=";
```

**Comment by SuperSandro2000** [comment]
@ofborg build kube-state-metrics

**Comment by eskytthe** [comment]
Sorry did not see you note about squash - my company use outlook - did hide the notification. But thanks for your help!

**Pull request merged**